src/Adapter/Customer/QueryHandler/GetCustomerInformationHandler.php

```diff
@@ -0,0 +1,590 @@
+        }
+
+        return new OrdersInformation(
+            Tools::displayPrice($totalSpent),
+            $validOrders,
+            $invalidOrders
+        );
+    }
+
+    /**
+     * @param Customer $customer
+     *
+     * @return CartInformation[]
+     */
+    private function getCustomerCarts(Customer $customer)
+    {
+        $carts = Cart::getCustomerCarts($customer->id);
+        $customerCarts = [];
+
+        foreach ($carts as $cart) {
+            $cart = new Cart((int) $cart['id_cart']);
+            Context::getContext()->cart = $cart;
+
+            $currency = new Currency($cart->id_currency);
+            Context::getContext()->currency = $currency;
+
+            $carrier = new Carrier($cart->id_carrier);
+            $summary = $cart->getSummaryDetails();
+
+            $customerCarts[] = new CartInformation(
+                sprintf('%06d', $cart->id),
+                Tools::displayDate($cart->date_add, null, true),
+                Tools::displayPrice($summary['total_price'], $currency),
+                $carrier->name
+            );
+        }
+
+        Context::getContext()->currency = Currency::getDefaultCurrency();
+
+        return $customerCarts;
+    }
+
+    /**
+     * @param Customer $customer
+     *
+     * @return ProductsInformation
+     */
+    private function getCustomerProducts(Customer $customer)
+    {
+        $boughtProducts = [];
+        $viewedProducts = [];
+
+        $products = $customer->getBoughtProducts();
+        foreach ($products as $product) {
+            $boughtProducts[] = new BoughtProductInformation(
+                (int) $product['id_order'],
+                Tools::displayDate($product['date_add'], null, false),
+                $product['product_name'],
+                $product['product_quantity']
+            );
+        }
+
+        $sql = '
+            SELECT DISTINCT cp.id_product, c.id_cart, c.id_shop, cp.id_shop AS cp_id_shop
+            FROM ' . _DB_PREFIX_ . 'cart_product cp
+            JOIN ' . _DB_PREFIX_ . 'cart c ON (c.id_cart = cp.id_cart)
+            JOIN ' . _DB_PREFIX_ . 'product p ON (cp.id_product = p.id_product)
+            WHERE c.id_customer = ' . (int) $customer->id . '
+                AND NOT EXISTS (
+                        SELECT 1
+                        FROM ' . _DB_PREFIX_ . 'orders o
+                        JOIN ' . _DB_PREFIX_ . 'order_detail od ON (o.id_order = od.id_order)
+                        WHERE product_id = cp.id_product AND o.valid = 1 AND o.id_customer = ' . (int) $customer->id . '
+                )
+        ';
+
+        $viewedProductsData = Db::getInstance()->executeS($sql);
+        foreach ($viewedProductsData as $productData) {
+            $product = new Product(
+                $productData['id_product'],
+                false,
+                $this->contextLangId,
+                $productData['id_shop']
+            );
+
+            if (!Validate::isLoadedObject($product)) {
+                continue;
```

Comment:
Should we `continue` ... or display a warning message (although I dont know how - exception ? flash message ?) ? It might be a PM decision.

Reply:
i think not a flash message for sure, because command can be dispatched using console application as well.

Reply:
Since this is a view, I would not want it to crash because of a bad product.
But we can log the issue at least :) so that it's easier to debug if someone is wondering "why my product X is not displayed in my customer view ?"

Reply:
hmm, i think it may be done this way on purpose.

if admin deletes a product that was previously viewed by customers, i dont think he cares any more about it, right? so why would you want to have history of all products (even those who were deleted year or more ago) that customer viewed? wdyt?